lib/dry/result_matcher/either_matcher.rb

```diff
@@ -0,0 +1,28 @@
+require "dry/result_matcher/case"
+
+module Dry
+  module ResultMatcher
+    EitherMatcher = Dry::ResultMatcher::Matcher.new(
+      success: Case.new(
+        match: -> result, *pattern {
+          result = result.to_either if result.respond_to?(:to_either)
+          result.right?
+        },
+        resolve: -> result {
+          result = result.to_either if result.respond_to?(:to_either)
+          result.value
+        },
+      ),
+      failure: Case.new(
+        match: -> result, *pattern {
+          result = result.to_either if result.respond_to?(:to_either)
+          result.left?
+        },
+        resolve: -> result {
+          result = result.to_either if result.respond_to?(:to_either)
+          result.value
+        },
+      )
```

Comment:
Can't we just assume it's an interface that we rely on? Try's success/failure could implement to_either. Would be cleaner.